tools/chpldef/chpldef.cpp

```diff
@@ -89,27 +93,35 @@ int main(int argc, char** argv) {
     CHPL_ASSERT(msg->status() == Message::PENDING);
     CHPL_ASSERT(!msg->isOutbound());
 
-    auto optRsp = Message::handle(ctx, msg.get());
-
-    // Always expect an immediate response for now.
-    if (!optRsp && !msg->isNotification()) {
-      CHPLDEF_FATAL(ctx, "Handler response should not be delayed!");
+    if (msg->tag() == Message::Exit) {
+      CHPL_ASSERT(ctx->state() == Server::SHUTDOWN);
```

Comment:
Not following. If you have _just received_ an Exit message, you must already have shutdown? Does something about `Message::request` affect the context state? If it does, that's very unintuitive.

I don't think `::compute` has happened yet.

Reply:
The protocol has the client send a `Shutdown` request to the server. The `Shutdown::compute` puts the server into the `SHUTDOWN` state.

We are doing the bare minimum to ensure correctness right now (asserts), but there are rules about what the server is allowed to do depending on what state of the handshake it is in.

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#shutdown

`Message::request` does not affect context state. Only `compute` methods will affect context state.